### DISCUSSION TITLE
Any way to find a players rank?

### DISCUSSION THREAD

**Discussion by TheFlyingVoid** [saul/demofile#215] [Questions] [answered]
I have been looking through the docs and I can't seem to find anything leading to players ranks. Is there any way of finding a players rank and if not is it possible to implement in the future?

**Comment by saul** [comment]
I'm not sure it is possible - the ranks are sent via Steam network channels and not the in-game network channels, so it's not recorded in the demo.

**Comment by luastoned** [comment]
What about the CCSUsrMsg_ServerRankUpdate protobuf?

**Comment by saul** [comment]
Potentially... I think if you listen for that message, you should be able to find the player whose rank has changed by comparing xuid_lo from PlayerInfo with account_id from the message.

Let me know how you get on.

**Comment by saul** [comment]
Will assume that this resolved your issue

**Comment by luastoned** [comment]
I don't think the issue is resolved (as noone has replied with a patch or working code).
How would one go about listening for that message anyway? I don't see any net related things exposed?

**Comment by saul** [comment]
```javascript
    demoFile.userMessages.on('ServerRankUpdate', um => {
      console.log('Player rank updates:');
      for (let update of um.rankUpdate) {
        let player = demoFile.players.find(player => player.userInfo.xuid.getLowBits() == update.accountId);
        console.log(`> ${player.name}: ${update.rankOld} -> ${update.rankNew}`);
      }
    });
```

**Comment by thecatontheflat** [comment]
Doesn't seem to work. Does it mean this message is not always sent?

**Comment by master117** [comment]
The rank is only sent at the end of MM matches, and only those recorded by gotv.

**Comment by thecatontheflat** [comment]
Thank you for the reply! I do check the MM match recorded by GOTV though.

http://replay192.valve.net/730/003377779566117388614_0320438711.dem.bz2

**Comment by saul** [comment, marked as answer]
I think the API has changed slightly since I posted that comment. See the below code (tested on your demo and works):

```ts
    demoFile.userMessages.on("ServerRankUpdate", um => {
      console.log("Player rank updates:");
      for (const update of um.rankUpdate) {
        const player = demoFile.players.find(
          player =>
            player.userInfo &&
            player.userInfo.xuid.getLowBits() === update.accountId
        );
        if (!player) console.log(`> (unknown player ${update.accountId})`);
        else
          console.log(
            `> ${player.name}: ${update.rankOld} -> ${update.rankNew}`
          );
      }
    });
```

Outputs:

```
Player rank updates:
> GundelGaukeley2: 11 -> 11   
> Frezzz chlenix.exe: 10 -> 10
> ็: 10 -> 10
> Cerberus: 9 -> 9
> d0h: 8 -> 8
> Comrade Gamer: 10 -> 10     
> Tor: 10 -> 10
> twitch.tv/jacus_minecraft_cs: 10 -> 10
> Łukasz Borzym: 9 -> 10
> (unknown player 396707352)
```
